scripts/check-components

```diff
@@ -0,0 +1,258 @@
+
+    # Check EventsSent events (common)
+    if @name.end_with? 'EventsSent'
+      members_must_include(['count', 'bytes_sent'])
+      counters_must_include(['sent_events_total', 'sent_event_bytes_total'])
+    end
+
+    # Check BytesSent events (for sinks)
+    if @name.end_with? 'BytesSent'
+      members_must_include(['byte_size'])
+    end
+
+    has_errors = @logs.one? { |type, _, _| type == 'error' }
+
+    # Make sure Error events output an error
+    if has_errors or @name.end_with? 'Error'
+      append('Error events MUST be named "___Error".') unless @name.end_with? 'Error'
+      counters_must_include(['errors_total'])
+    end
+
+    # Make sure error events contain the right parameters
+    @logs.each { |type, message, parameters|
+      if type == 'error'
+        ['error', 'stage'].each { |parameter|
+          unless parameters.include? parameter
+            @reports.append("Error log MUST include parameter \"#{parameter}\".")
+          end
+        }
+      end
+    }
+
+    unless @reports.empty?
+      puts "#{@path}: Errors in event #{@name}:"
+      @reports.each { |report|
+        puts "    #{report}"
+        $error_count += 1
+      }
+    end
+  end
+
+  private
+
+  def append(report)
+    @reports.append(report)
+  end
+
+  def generic_must_contain(array, names, prefix, suffix)
+    names.each { |name|
+      unless array.include? name
+        @reports.append("#{prefix} MUST #{suffix} \"#{name}\".")
+      end
+    }
+  end
+
+  def counters_must_include(names)
+    generic_must_contain(@counters, names, 'This event', 'increment counter')
+    # TODO: Scan each counter for parameters
+  end
+
+  def members_must_include(names)
+    generic_must_contain(@members, names, 'This event', 'have a member named')
+  end
+end
+
+class Component
+  def initialize(name)
+    @name = name
+    @emits = []
+  end
+
+  def emits(array)
+    @emits += array
+  end
+
+  def validate
+    @missing = []
+    require_events(SOURCE_EVENTS) if @name.start_with? 'sources/'
+    require_events(TRANSFORM_EVENTS) if @name.start_with? 'transforms/'
+    require_events(SINK_EVENTS) if @name.start_with? 'sinks/'
+
+    if @missing.length > 0
+      puts "#{@name}: Missing required event:"
+      @missing.each { |event|
+        puts "    #{event}"
+        $error_count += 1
+      }
+    end
+  end
+
+  private
+
+  def require_events(names)
+    names.each { |name|
+      if @emits.none? { |event| event.end_with? name }
+        @missing.append(name)
+      end
+    }
+  end
+end
+
+$all_events = Hash::new { |hash, key| hash[key] = Event::new(key) }
+$all_components = Hash::new { |hash, key| hash[key] = Component::new(key) }
+
+# Scan sources and build internal structures
+Find.find('src') { |path|
+  if path.end_with? '.rs'
+    text = File.read(path)
+
+    # Check log message texts for correct formatting. See below for the
+    # full regex
+    text.scan(/(trace|debug|info|warn|error)!\(\s*(message\s*=\s*)?"([^({)][^("]+)"/) {
```

Comment:
nit: I'd probably build a class to contain all of this logic so that the loop is simplified.